sdk/python/lib/pulumi/runtime/rpc_manager.py

```diff
@@ -65,11 +65,13 @@ def do_rpc(self, name: str, rpc_function: Callable[..., Awaitable[None]]) -> Cal
                 log.debug(f"recorded new RPC, {self.count} RPCs outstanding")
 
             try:
-                await rpc_function(*args, **kwargs)
+                result = await rpc_function(*args, **kwargs)
             except Exception as exn:
                 log.debug(f"RPC failed with exception:")
                 log.debug(traceback.format_exc())
-                self.exception_future.set_exception(exn)
+                if not self.exception_future.done():
+                    self.exception_future.set_exception(exn)
```

Comment:
If multiple coroutines wait on a single future and that future terminates with an exception, that exception can be thrown multiple times. If we have multiple RPCs in-flight waiting on that future, each one of those RPCs may receive the same exception.

This fix only signals the `exception_future` if it hasn't been signaled already, because `set_exception` will throw if it has already been called.